### PR TITLE
Copy program support

### PR DESCRIPTION
**Pull request by scxiao**
Add the copy constructor and copy assignment operator for program..

### CODE REVIEW

File: src/program.cpp [hunk 1]
```diff
@@ -89,6 +89,73 @@ program::program(program&&) noexcept = default;
...
 
+// copy constructor
+program::program(const program& p) { copy(p); }
+
+// copy assignment operator
+program& program::operator=(const program& p)
+{
+    if(this != &p)
+    {
+        copy(p);
```
Comment (pfultz2) [inline src/program.cpp]: The assignment operator should use Copy-Swap:

```cpp
program& program::operator=(program p)
{
    std::swap(p.impl, impl);
}
```

File: src/program.cpp [hunk 2]
```diff
@@ -89,6 +89,73 @@ program::program(program&&) noexcept = default;
...
+            copy_ins = impl->instructions.insert(impl->instructions.end(), instruction{l});
+        }
+        else if(ins->name() == "@param")
+        {
+            auto&& name = any_cast<builtin::param>(ins->get_operator()).parameter;
+            auto s      = ins->get_shape();
+            copy_ins    = impl->instructions.insert(impl->instructions.end(),
+                                                 {builtin::param{name}, std::move(s), {}});
+        }
+        else if(ins->name() == "@outline")
```
Comment (pfultz2) [inline src/program.cpp]: Do we need this?

File: test/program_test.cpp [hunk 3]
```diff
@@ -27,4 +30,46 @@ TEST_CASE(program_equality)
...
+        p.add_instruction(migraphx::op::mul{}, sum, po);
+
+        return p;
+    };
+
+    {
+        auto p1 = create_program_1();
+        auto p2 = p1;
+        p2.compile(migraphx::cpu::target{});
+        EXPECT(p1 == p2);
```
Comment (pfultz2) [inline test/program_test.cpp]: We should have a test that `p1 != p2` after compiling.

Comment (pfultz2) [reply test/program_test.cpp]: But this should be `false` since `op::add` will get lowered to `cpu::add`.

File: src/program.cpp [hunk 4]
```diff
@@ -86,8 +86,70 @@ static void print_program(const program& p, F print_func)
...
+program::program(const program& p) { copy(p); }
+
+// copy assignment operator
+program& program::operator=(program p)
+{
+    std::swap(p.impl, this->impl);
+    return *this;
+}
+
+void program::copy(const program& p)
```
Comment (pfultz2) [inline src/program.cpp]: This should be named `assign` not `copy`.

File: src/program.cpp [hunk 5]
```diff
@@ -86,8 +86,70 @@ static void print_program(const program& p, F print_func)
...
+{
+    // clean the current program
+    if(!impl)
+    {
+        impl = std::make_unique<program_impl>();
+    }
+    else if(!impl->instructions.empty())
+    {
+        remove_instructions(begin(), end());
```
Comment (pfultz2) [inline src/program.cpp]: Is it possible to call `impl->instructions.clear()` here instead?

File: src/program.cpp (hunk 1 above)
Comment (scxiao) [reply src/program.cpp]: fixed

File: test/program_test.cpp (hunk 3 above)
Comment (scxiao) [reply test/program_test.cpp]: rename the name of the cpu operator, now it works as expected.

File: src/program.cpp (hunk 4 above)
Comment (scxiao) [reply src/program.cpp]: fixed

File: src/program.cpp (hunk 5 above)
Comment (scxiao) [reply src/program.cpp]: I think it is possible and changed

File: src/program.cpp (hunk 2 above)
Comment (scxiao) [reply src/program.cpp]: I have no idea, I checked the function generic_eval(), which processing @outline instructions.

Comment (pfultz2) [reply src/program.cpp]: I think we are going to remove `@outline` eventually. We dont need it anymore. I guess it can stay there for now.

**Comment by mvermeulen** [comment]
@mei-ye review?

File: src/program.cpp [hunk 6]
```diff
@@ -86,8 +86,70 @@ static void print_program(const program& p, F print_func)
...
+            // retrieve its mapped input
+            auto inputs = ins->inputs();
+            // ensure all inputs have its corresponding copy instructions
+            assert(std::all_of(
+                inputs.begin(), inputs.end(), [&](auto i) { return ins_map.count(i) > 0; }));
+            std::vector<instruction_ref> copy_inputs(inputs.size());
+            std::transform(inputs.begin(), inputs.end(), copy_inputs.begin(), [&](auto i) {
+                return ins_map[i];
+            });
+            copy_ins = add_instruction(ins->get_operator(), copy_inputs);
```
Comment (mei-ye) [inline src/program.cpp]: Sorry for the late reply.  A question: is it sufficient to copy operator and inputs?   An operation might have its own attribute fields (e.g., cached CPU op in GPU op), do we need to copy those attributes?  Alternatively, this question can be asked in this way:  is our op design general enough to allow copies?  If not, each operator needs to have its own copy operators.

Comment (scxiao) [reply src/program.cpp]: Thanks for pointing this out. I tested that offline by copying programs imported from onnx files. The default copy constructor of operators will copy their attributes, and I added one more tests for that.

Comment (mei-ye) [reply src/program.cpp]: So this program copy is meant to be used for copying programs created from ONNX?
If instructions are lowered into GPU instructions, the copy mechanism might not work?  Iff this is what it is designed for, then it is OK with me.

Comment (scxiao) [reply src/program.cpp]: the lowered program (in both CPU and GPU) can also be copied correctly. In quantization, we mainly do program copy before compilation.

**Pull request merged**